runtime/datetime.c

```diff
@@ -1258,6 +1260,64 @@ timeConvertToUTC(const struct syslogTime *const __restrict__ local,
 	timeval2syslogTime(&tp, utc, 1);
 }
 
+/**
+ * Format a UNIX timestamp.
+ */
+static int
+formatUnixTimeFromTime_t(time_t unixtime, const char *format, char *pBuf, uint pBufMax) {
+	struct tm lt;
+
+	assert(format != NULL);
+	assert(pBuf != NULL);
+
+	// Let's not depend on debug asserts to save the day in the wild!
```

Comment:
I would **not** do this in non-debug build. After all, that's a calling sequence error which should never happen. If we do checks all over the block, this is pretty time consuming.

Reply:
I come from a security background, and worry about CVEs. Where things that should never happen do happen :)

This situation would never occur in the existing calling sequence, but what about if/when this function is called from somewhere else in the future?

Reply:
I see the point. But if we introduce all those checks consistently, we will loose quite a lot of performance. I think there needs to be made a compromise. Maybe no longer, but than the concept of high performance also needs to be redefined.

If the calling sequence is violated, it's a bug in any case. Again, I see the dilemma here. But if we really want to handle that, than we need to emit a very big user error message, because a calling sequence violation is a very strong indication that something got really, really wrong. Doing it just one place doesn't make sense IMHO, so we would need to do it at the other 2,000 or so places as well (not exaggegrating). @davidelang what do you think?

Reply:
I suppose there's really no way the configurer/user could somehow use a value that would result in a NULL being passed to the function. Also, NULL checks are rather superficial anyway - If the pointer were somehow set to an address of 0x00000001 (just as likely as being accidentally set to 0x00000000) rsyslog would still segfault.

Reply:
The NULL checks actually date back to pre-valgrind and pre-ASAN. I often remove them now nowadays as either of the two will catch them even better as assert(). I do use assert still for some invariant checks that point to logic errors. And this is my point: if we intend to handle all this with more checks and workarounds, I think in the long term we get "needless" complex code that probably results in more CVEs (due to logic errors) than when we accept that we cannot cover everything (and let the hardening done by the compiler as much as possible). Of course, proper testing is important here as well. Thankfully we've come a long way on the CI front compared to e.g. 5 years ago. Still more work to be done (I don't get unit tests going due to problems with linking that nobody ever managed to tell me how to do it right :-( ).